### PR TITLE
PROGRAMMES-6141 Paginator

### PR DESCRIPTION
**Pull request by votemike**
https://jira.dev.bbc.co.uk/browse/PROGRAMMES-6141

Can be tested on programmes/b006qykl/episodes/player

This is similar to the paginator used in /blogs. Would it be worth making a composer package that both of them can use? Or not worth it?

### CODE REVIEW

**Comment by programmes-developers** [comment]
Can one of the admins verify this patch?

File: src/Ds2013/Presenters/Utilities/Paginator/PaginatorPresenter.php [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    {
+        return (int) ceil($this->totalItems / $this->pageSize);
+    }
+
+    private function buildItems(): array
+    {
+        $pages = $this->getPageCount();
+        if ($pages <= 7 || ($pages === 8 && ($this->currentPage === 4 || $this->currentPage === 5)) || ($pages === 9 && $this->currentPage === 5)) {
+            $items = [];
+            for ($page = 1; $page <= $pages; $page++) {
```
Comment (felipeparaujo) [inline src/Ds2013/Presenters/Utilities/Paginator/PaginatorPresenter.php]: everyone knows the right to do this is `$page = 0; $page < $pages; $page++`

File: src/Ds2013/Presenters/Utilities/Paginator/PaginatorPresenter.php [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+    {
+        return abs($item - $this->currentPage);
+    }
+
+    public function getPageCount(): int
+    {
+        return (int) ceil($this->totalItems / $this->pageSize);
+    }
+
+    private function buildItems(): array
```
Comment (felipeparaujo) [inline src/Ds2013/Presenters/Utilities/Paginator/PaginatorPresenter.php]: Please add a lot of comments explaining what's going on. This is suuuper weird

File: src/Ds2013/Presenters/Utilities/Paginator/paginator.html.twig [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    {%- else -%}
+        <li class="pagination__previous">
+            <a href="?page={{ paginator.getCurrentPage() - 1 }}" rel="prev">
+                {{ gelicon('core', 'previous', 'pagination__next-previous gelicon--leading') }}
+                <span>{{ tr('previous') }}</span>
+            </a>
+        </li>
+    {%- endif -%}
+
+    {% for item in paginator.getItems() %}
```
Comment (felipeparaujo) [inline src/Ds2013/Presenters/Utilities/Paginator/paginator.html.twig]: `{%-` ?

File: src/Ds2013/Presenters/Utilities/Paginator/paginator.html.twig [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+        {%- if item == 'spacer' -%}
+            <li class="pagination__spacer">
+                <span>&#8230;</span>
+            </li>
+        {%- else -%}
+            {%- set classes = build_css_classes({
+                'pagination__page': true,
+                ('pagination__page--offset' ~ paginator.getOffset(item)): true,
```
Comment (felipeparaujo) [inline src/Ds2013/Presenters/Utilities/Paginator/paginator.html.twig]: only `pagination__page--offset1` and `pagination__page--offset2` are defined. Is this correct?

File: src/Ds2013/Presenters/Utilities/Paginator/PaginatorPresenter.php (hunk 2 above)
Comment (votemike) [reply src/Ds2013/Presenters/Utilities/Paginator/PaginatorPresenter.php]: The tests explain it all...

File: src/Ds2013/Presenters/Utilities/Paginator/PaginatorPresenter.php (hunk 1 above)
Comment (votemike) [reply src/Ds2013/Presenters/Utilities/Paginator/PaginatorPresenter.php]: 🤔

File: src/Ds2013/Presenters/Utilities/Paginator/paginator.html.twig (hunk 4 above)
Comment (votemike) [reply src/Ds2013/Presenters/Utilities/Paginator/paginator.html.twig]: Should be fine. This is to do with how many pages either side of the current page are shown at smaller screen sizes.

File: src/Ds2013/Presenters/Utilities/Paginator/PaginatorPresenter.php (hunk 2 above)
Comment (felipeparaujo) [reply src/Ds2013/Presenters/Utilities/Paginator/PaginatorPresenter.php]: I shouldn't have to go to the tests to understand what the code does...

File: src/Ds2013/_ds2013-components.scss [hunk 5]
```diff
@@ -68,6 +68,7 @@
 @import 'Css/Components/box-link';
 @import 'Presenters/Utilities/block-link';
 @import 'Presenters/Utilities/lazyload';
+@import 'Presenters/Utilities/Paginator/paginator';
```
Comment (BPScott) [inline src/Ds2013/_ds2013-components.scss]: Delete the `@import 'objects/pagination';` above.

**Pull request merged**